File: scripts/i18n/seed-api-translations.py
#!/usr/bin/env python3
"""
Translate API message keys (lang/en.json) into all supported locales.
Uses deep-translator (Google Translate, no API key).
"""

from __future__ import annotations

import json
import sys
import time
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
EN_PATH = ROOT / "lang" / "en.json"

LOCALES = {
    "bn": "bn",
    "hi": "hi",
    "ar": "ar",
    "zh": "zh-CN",
    "ur": "ur",
    "fa": "fa",
    "tr": "tr",
    "es": "es",
    "it": "it",
}

try:
    from deep_translator import GoogleTranslator
except ImportError:
    print("Installing deep-translator...", file=sys.stderr)
    import subprocess

    subprocess.check_call([sys.executable, "-m", "pip", "install", "deep-translator", "-q"])
    from deep_translator import GoogleTranslator


def translate_text(text: str, target: str, cache: dict[tuple[str, str], str]) -> str:
    key = (target, text)
    if key in cache:
        return cache[key]
    if not text.strip():
        cache[key] = text
        return text
    try:
        translated = GoogleTranslator(source="en", target=target).translate(text)
    except Exception as exc:  # noqa: BLE001
        print(f"  warn: {exc!r} for {text[:40]!r}", file=sys.stderr)
        translated = text
    cache[key] = translated
    time.sleep(0.08)
    return translated


def main() -> None:
    en: dict[str, str] = json.loads(EN_PATH.read_text(encoding="utf-8"))
    cache: dict[tuple[str, str], str] = {}

    for file_locale, google_target in LOCALES.items():
        out_path = ROOT / "lang" / f"{file_locale}.json"
        existing: dict[str, str] = {}
        if out_path.exists():
            existing = json.loads(out_path.read_text(encoding="utf-8"))

        print(f"Translating API messages -> {file_locale} ({len(en)} keys)...")
        translated: dict[str, str] = {}
        for i, (msg_key, _en_val) in enumerate(en.items(), 1):
            if (
                file_locale == "bn"
                and msg_key in existing
                and existing[msg_key] != msg_key
                and existing[msg_key] != en.get(msg_key)
            ):
                translated[msg_key] = existing[msg_key]
                continue
            translated[msg_key] = translate_text(msg_key, google_target, cache)
            if i % 25 == 0:
                print(f"  {file_locale}: {i}/{len(en)}")

        out_path.write_text(
            json.dumps(translated, ensure_ascii=False, indent=4, sort_keys=True) + "\n",
            encoding="utf-8",
        )
        print(f"  wrote {out_path}")

    print("API lang/*.json complete.")


if __name__ == "__main__":
    main()

File: scripts/i18n/seed-api-translations.php
#!/usr/bin/env php
<?php

declare(strict_types=1);

/**
 * Translate lang/en.json into all supported API locales.
 */

require __DIR__.'/lib/translate.php';

$root = dirname(__DIR__, 2);
$enPath = $root.'/lang/en.json';
$en = json_decode(file_get_contents($enPath), true, 512, JSON_THROW_ON_ERROR);

$targets = [
    'bn' => 'bn',
    'hi' => 'hi',
    'ar' => 'ar',
    'zh' => 'zh-CN',
    'ur' => 'ur',
    'fa' => 'fa',
    'tr' => 'tr',
    'es' => 'es',
    'it' => 'it',
];

foreach ($targets as $fileLocale => $googleTarget) {
    $outPath = $root.'/lang/'.$fileLocale.'.json';
    $existing = is_file($outPath)
        ? json_decode(file_get_contents($outPath), true) ?: []
        : [];

    $alreadyDone = count($en) > 0 && count(array_filter(
        $en,
        static fn (string $key): bool => isset($existing[$key]) && $existing[$key] !== $key
    )) >= count($en);

    if ($alreadyDone) {
        echo "API -> {$fileLocale} skipped (already translated)\n";
        continue;
    }

    echo "API -> {$fileLocale} (".count($en)." keys)\n";
    $out = [];
    $i = 0;
    foreach ($en as $key => $_) {
        ++$i;
        if (
            $fileLocale === 'bn'
            && isset($existing[$key])
            && $existing[$key] !== $key
            && mb_strlen($existing[$key]) > 0
        ) {
            $out[$key] = $existing[$key];
            continue;
        }
        $out[$key] = i18n_translate_en($key, $googleTarget);
        if ($i % 30 === 0) {
            echo "  {$i}/".count($en)."\n";
        }
    }

    ksort($out);
    file_put_contents(
        $outPath,
        json_encode($out, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES)."\n"
    );
    echo "  wrote {$outPath}\n";
}

echo "API lang/*.json complete.\n";
